### PR TITLE
Add ec2 IPv6 IMDS Support (SC-336)

### PR DESCRIPTION
**Pull request by holmanb**
```
Add support for dual stack ipv6/ipv4 IMDS to Ec2

- add support for parallel http(s) requests to wait_for_url()
- implementation based loosely on RFC6555: "Happy Eyeballs" [1]
- update ec2 datasource to support dual-stack ipv6/ipv4
- only "nitro" instances have ipv6 IMDS, favor ipv4 initially
- replace httpretty with requests for ec2 tests

[1] https://datatracker.ietf.org/doc/html/rfc6555
```

## Additional Context

In contrast to "happy eyeballs", this implementation supports parallel attempts to communicate with a small delay favoring ipv4. This is datasource-specific, and other datasources that wish favor ipv6 can do so trivially.

### Technical Description of Changes:
1.  use a threadpool executor to wrap synchronous calls in async scheduler (see `url_helper.py:dual_stack()`) that does 0.150s delay, cancels pending responses, etc (basically a stateless generic happy eyeballs implementation for synchronous libraries but at the http layer, not the socket/connection layer)
2. refactor `wait_for_url()` into reusable pieces to support optionally parallel http requests to endpoints
3. add tests that verify behavior by simulating a slow/unresponsive endpoint
4. update Ec2 datasource to support parallel GETs (PUTs are left synchronous)
5. update Ec2 tests to test with ipv6
6. make Ec2 favor ipv4 (for now), when majority of cloud has ipv6 support this can change to favor ipv6

### CODE REVIEW

**Comment by holmanb** [comment]
I'm currently tracking down an issue where tests pass with httpretty when ran directly, but when running pytest locally against all of cloud-init, they fail [1].

I'm fairly confident that the behavior those tests are supposed to be testing are not the cause for this issue. Also the tests pass via travis, so I switched this from draft to ready to review.

[1] The failure message is: `No mocking was registered, and real connections are not allowed (httpretty.allow_net_connect = False).`, which is odd, given that these tests _are_ mocked using `httpretty.register_uri()` which works correctly unless ran with the rest of the suite.

File: cloudinit/url_helper.py [hunk 1]
```diff
@@ -442,18 +516,82 @@ def timeup(max_wait, start_time):
...
+
+    def read_url_serial():
+        """iterate over list of urls, request each one and handle responses
+        and thrown exceptions individually per url
+        """
+
+        def url_reader_serial(url):
+            return (url, read_url_cb(url))
+
+        nonlocal timeout
```
Comment (TheRealFalcon) [inline cloudinit/url_helper.py:588]: I think the nonlocal here is a smell. If I follow it through, there's nothing dangerous about it, but it'd be easier to reason about if timeout was passed to the function and then passed again to `read_url_cb`.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Thanks @holmanb . This has come together really nicely. I have just a few small comments.

Do we need to change the call on line 250 of DataSourceEc2.py to be asynchronous as well?
I think the `start_time` and `status_cb` assignments in `wait_for_url` might be better placed under all of the nested functions.

I went back and forth about how abstract this is. On one hand, the code is elegant and well structured. On the other hand, it can be hard to follow the flow through the layers of callbacks, partials, and closures. But then again, unwinding it all or adding types might cause it to be even harder to follow. I think at the end of the day, this is just a more functional approach, and it's really just a few functions, so I think it's fine.

Have you tested this on a live ec2 instance yet? I think it'd be worth ensuring we can fetch metadata in both ipv4-only and ipv6-only environments. If not, we should do that before we merge this. Can you post your test steps in the PR description so they can be easily reproduced in the future? 

I think it'd be worth figuring out how to instrument pycloudlib to launch an ipv6 instance so we can write an integration test, but that doesn't need to come with this PR...though it'd be nice before our next SRU.

File: cloudinit/url_helper.py [hunk 2]
```diff
@@ -382,6 +385,73 @@ def _cb(url):
...
+        return func(addr)
+
+    executor = ThreadPoolExecutor(max_workers=len(addresses))
+    try:
+        futures = {
+            executor.submit(
+                _run_func,
+                func=func,
+                addr=addr,
+                delay=(None if i == 0 else stagger_delay),
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/url_helper.py:417]: Good catch! That logic comes from an early iteration when dual_stack() supported only 2 addresses (per RFC 6555). I missed that when I extended this function to support >2 addresses.

File: cloudinit/url_helper.py (hunk 1 above)
Comment (holmanb) [reply cloudinit/url_helper.py:588]: Sounds good to me

**Comment by holmanb** [comment]
>This has come together really nicely.

Thank you :)

> Do we need to change the call on line 250 of DataSourceEc2.py to be asynchronous as well?

I debated this myself, and honestly I still don't know the right answer. I need a better understanding of the semantics of AWS tokens. I should have noted this in the comments when I switched to requesting review, apologies.

I do know that if we do make the PUT on line 250 async, the assertion at `test_aws_token_redacted():673` will fail with len == 5, because the PUT happens multiple times. What does this actually mean?

> I think the `start_time` and `status_cb` assignments in `wait_for_url` might be better placed under all of the nested functions.

Agreed
 
> I went back and forth about how abstract this is. On one hand, the code is elegant and well structured. On the other hand, it can be hard to follow the flow through the layers of callbacks, partials, and closures. But then again, unwinding it all or adding types might cause it to be even harder to follow. I think at the end of the day, this is just a more functional approach, and it's really just a few functions, so I think it's fine.

My sentiments exactly. I should note that since this code already had lots of callbacks it didn't seem too out of place when writing it. That said, I can always restructure to something less functional if needed. One concern I have with doing that is that there would likely be more code duplication and resulting potential for subtle differences in behavior between the sync/parallel implementations. 

> Have you tested this on a live ec2 instance yet? I think it'd be worth ensuring we can fetch metadata in both ipv4-only and ipv6-only environments. If not, we should do that before we merge this. Can you post your test steps in the PR description so they can be easily reproduced in the future?

I have not yet. Since I haven't heard too strong of objections on this yet I'll get started on that next.

> I think it'd be worth figuring out how to instrument pycloudlib to launch an ipv6 instance so we can write an integration test, but that doesn't need to come with this PR...though it'd be nice before our next SRU.

I like that idea.

File: cloudinit/sources/DataSourceEc2.py [hunk 3]
```diff
@@ -317,6 +321,7 @@ def wait_for_metadata_service(self):
                 headers_redact=AWS_TOKEN_REDACT,
                 headers_cb=self._get_headers,
                 request_method=request_method,
+                connect_synchronously=False,
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceEc2.py:324]: On EC2 official platform (which is the only Ec2 lookalike that we are currently aware of with IPv6 IMDS support) we should be providing  this connect_synchronously=False only to the call `uhelp.wait_for_url` in `_maybe_fetch_api_token` on line 259.  We shouldn't attempt this here as our first interaction to official EC2 IMDS platform is the latest/api/token which is an "allowed" URL for us to hit/retry multiple times without triggering "non-token IMDS interaction warnings/counts" on EC2 instances.

Comment (blackboxsw) [reply cloudinit/sources/DataSourceEc2.py:324]: And given that `_maybe_fetch_api_token` is only intended to support EC2 proper platforms, this feels like the right scope for this changeset anyway. If other EC2-lookaline platforms grow IPv6 IMDS support we can adapt in the future.

**Review by blackboxsw — COMMENTED** [review]
This is great work. Please add a fairly detailed suggested commit message in the description of the PR as it merits a lot more context about the design/architecture here.
Also a pointer to the  RFC 6555 link in the commit message will be helpful.

File: cloudinit/url_helper.py [hunk 4]
```diff
@@ -382,6 +385,74 @@ def _cb(url):
...
+        returned_address = futures[future]
+        return_result = future.result()
+        return_exception = future.exception()
+        if return_exception:
+            LOG.warning(
+                "Exception %s during request to %s",
+                return_exception,
+                returned_address,
+            )
+            raise return_exception
```
Comment (blackboxsw) [inline cloudinit/url_helper.py:435]: Would we want to specifically look at requests.ConnectionError failures which basically tell us no route to host for IPv6 and ignore them as no route means we really aren't ever going to succeed there right?

Comment (holmanb) [reply cloudinit/url_helper.py:435]: Interesting question. I don't think that "no route to host" should ever happen (because we're using link-local addresses[1]), but maybe I'm wrong. Generally the idea of "ignore connection failures in case the others succeed" is something I can get behind.

[1] To my knowledge Linux doesn't require fib entries for sending directly to a link-local addresses.

A quick experiment:
```
[ec2-user@ip-172-31-6-134 ~]$ ip r
default via 172.31.0.1 dev eth0 
169.254.169.254 dev eth0 
172.31.0.0/20 dev eth0 proto kernel scope link src 172.31.6.134 
[ec2-user@ip-172-31-6-134 ~]$ sudo ip r del 169.254.169.254
[ec2-user@ip-172-31-6-134 ~]$ ip route show table all | grep -e 169 -e ec2
[ec2-user@ip-172-31-6-134 ~]$ wget http://169.254.169.254
--2022-01-25 23:39:54--  http://169.254.169.254/
Connecting to 169.254.169.254:80... connected.
HTTP request sent, awaiting response... 200 OK
```
Even though the link local route was removed from the routing table it was still able to communicate using that address.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
forgot to set request changes for moving the connect_synchronously up into _maybe_fetch_api_token.

File: cloudinit/url_helper.py (hunk 4 above)
Comment (blackboxsw) [reply cloudinit/url_helper.py:435]: Ahh maybe I was misinterpreting the log I saw on an IPv4 only instance:
`2022-01-25 22:34:46,195 - url_helper.py[WARNING]: Calling 'None' failed [0/120s]: request error [HTTPConnectionPool(host='fd00:ec2::254', port=80): Max retries exceeded with url: /latest/api/token (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7facb9c98400>: Failed to establish a new connection: [Errno 101] Network is unreachable'))]`

Comment (holmanb) [reply cloudinit/url_helper.py:435]: Thanks for the context. "Network is unreachable" is often related to routing issues, I see what you were getting at. Regardless of what "Network is unreachable" is caused by,  I think this suggestion is a good one. Specifically this would make connectivity possible in the case that there is a slow first address and an unreachable second address.

File: cloudinit/url_helper.py [hunk 5]
```diff
@@ -382,6 +385,74 @@ def _cb(url):
...
+        elif not return_result:
+            LOG.warning("Empty result for address %s", returned_address)
+
+    # when max_wait expires
+    except TimeoutError:
+        LOG.warning("Timed out waiting for addresses: %s", " ".join(addresses))
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/url_helper.py:441]: This is thrown from `as_completed()` if none of the endpoints responds before the timeout. If one responds, this does not get thrown.

Snippet from the [docs](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.as_completed):
> The returned iterator raises a concurrent.futures.TimeoutError if `__next__()` is called and the result isn’t available after timeout seconds from the original call to as_completed().

 This is the same timeout passed to urlopen and for Ec2 this is currently set to 50.

File: cloudinit/url_helper.py [hunk 6]
```diff
@@ -382,6 +385,74 @@ def _cb(url):
...
+    # Executor doesn't provide kwargs for setting shutdown behavior
+    # in the constructor, otherwise the context manager would be preferred
+    finally:
+        # python 3.9 allows canceling futures, which may save some cycles
+        if sys.version_info.major >= 3 and sys.version_info.minor > 9:
+            executor.shutdown(  # pylint: disable=E1123
+                wait=False, cancel_futures=True
+            )
+        else:
+            executor.shutdown(wait=False)
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/url_helper.py:452]: Agreed, this is easier to visually parse.

**Comment by blackboxsw** [comment]
> I think it'd be worth figuring out how to instrument pycloudlib to launch an ipv6 instance so we can write an integration test, but that doesn't need to come with this PR...though it'd be nice before our next SRU.


pycloudlib does have capacity to launch dual-stack ipv4/ipv6 vpc (though this may not yet against exclusive IPv6 IMDS) via something like:
```python
vpc = ec2.get_or_create_vpc(name="cloud-nit-dualstack-vpc")
ec2.launch(image_id, vpc=vpc)
```

So I think we can adapt subclass  tests.integrationtests.cloud.Ec2Cloud._perform_launch to call the get_or_create_vpc  first in cases we want dual-stack support for testing

File: cloudinit/url_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/url_helper.py]: test added in `tests/unittests/test_url_helper.py:test_dual_stack_staggered()`

File: cloudinit/url_helper.py (hunk 4 above)
Comment (holmanb) [reply cloudinit/url_helper.py:434]: > Interesting question. I don't think that "no route to host" should ever happen (because we're using link-local addresses[1]), but maybe I'm wrong. 

Update: I was wrong.

My former argument here was invalid. The link local address used on ec2 instances is actually the destination address, not source address. Http(s) traffic leaves via the default route on a private address (RFC1918).

Remove default routes and this error could happen:

```
[ec2-user@ip-172-31-6-134 ~]$ wget http://[fd00:ec2::254]/latest/meta-data/
--2022-02-10 22:36:17--  http://[fd00:ec2::254]/latest/meta-data/
Connecting to [fd00:ec2::254]:80... failed: Network is unreachable.
```

I still think that the route I deleted in the example above (`169.254.169.254 dev eth0`) is unneeded, but for different reasons.

This comment doesn't imply a different course of action, but I wanted to set the record straight here.

**Comment by TheRealFalcon** [comment]
The failing test is an annoying bug that was fixed a few years ago:
https://github.com/getsentry/responses/issues/212

Here's a workaround:
```
diff --git a/tests/unittests/sources/test_ec2.py b/tests/unittests/sources/test_ec2.py
index 3848924d..9e373632 100644
--- a/tests/unittests/sources/test_ec2.py
+++ b/tests/unittests/sources/test_ec2.py
@@ -530,6 +530,14 @@ class TestEc2(test_helpers.HttprettyTestCase):
             md={"md": old_metadata},
         )
         self.assertTrue(ds.get_data())
+
+        # Workaround https://github.com/getsentry/responses/issues/212
+        if hasattr(responses.mock, "_urls"):
+            for index, url in enumerate(responses.mock._urls):
+                if url["url"].startswith(
+                    "http://169.254.169.254/2009-04-04/meta-data/"
+                ):
+                    del responses.mock._urls[index]
         # Provide new revision of metadata that contains network data
         register_mock_metaserver(
             "http://169.254.169.254/2009-04-04/meta-data/", DEFAULT_METADATA
```

Since it's only one test, I'm fine just inserting it there. If this becomes a bigger issue we'd probably want to find a more elegant solution.

**Comment by paride** [comment]
@TheRealFalcon nice! Also matches my findings about responses 0.17.0 being the first version without this issue. Maybe worth adding a line to the comment explaining that the workaround can be removed once there will be no need for testing with requests < 0.17.0 (that is, post-Focal EOL)?

**Comment by holmanb** [comment]
> The failing test is an annoying bug that was fixed a few years ago: [getsentry/responses#212](https://github.com/getsentry/responses/issues/212)

Thanks! This works locally so I have included it.

Also, I had to rebase on the tip of main since the latest commit introduced a merge conflict in one of the tests I changed.

**Comment by holmanb** [comment]
I believe this is ready for re-review.

**Comment by TheRealFalcon** [comment]
I made an integration test for this. If you use https://github.com/canonical/pycloudlib/pull/188 and apply this patch, it should work
```diff
diff --git a/tests/integration_tests/clouds.py b/tests/integration_tests/clouds.py
index e5fe56d1..5ea41cbb 100644
--- a/tests/integration_tests/clouds.py
+++ b/tests/integration_tests/clouds.py
@@ -207,10 +207,14 @@ class Ec2Cloud(IntegrationCloud):
 
     def _perform_launch(self, launch_kwargs, **kwargs):
         """Use a dual-stack VPC for cloud-init integration testing."""
-        launch_kwargs["vpc"] = self.cloud_instance.get_or_create_vpc(
-            name="ec2-cloud-init-integration"
-        )
+        if "vpc" not in launch_kwargs:
+            launch_kwargs["vpc"] = self.cloud_instance.get_or_create_vpc(
+                name="ec2-cloud-init-integration"
+            )
+        if "Ipv6AddressCount" not in launch_kwargs:
+            launch_kwargs["Ipv6AddressCount"] = 1
         pycloudlib_instance = self.cloud_instance.launch(**launch_kwargs)
+        pycloudlib_instance.enable_ipv6_instance_metadata()
         return pycloudlib_instance
 
 
diff --git a/tests/integration_tests/datasources/test_ec2.py b/tests/integration_tests/datasources/test_ec2.py
new file mode 100644
index 00000000..8cde4dc9
--- /dev/null
+++ b/tests/integration_tests/datasources/test_ec2.py
@@ -0,0 +1,43 @@
+import re
+
+import pytest
+
+from tests.integration_tests.instances import IntegrationInstance
+
+
+def _test_crawl(client, ip):
+    assert client.execute("cloud-init clean --logs").ok
+    assert client.execute("cloud-init init --local").ok
+    log = client.read_from_file("/var/log/cloud-init.log")
+    assert f"Using metadata source: '{ip}'" in log
+    result = re.findall(
+        r"Crawl of metadata service took (\d+.\d+) seconds", log
+    )
+    if len(result) != 1:
+        pytest.fail(f"Expected 1 metadata crawl time, got {result}")
+    # 20 would still be a crazy long time for metadata service to crawl,
+    # but it's short enough to know we're not waiting for a response
+    assert float(result[0]) < 20
+
+
+@pytest.mark.ec2
+def test_dual_stack(client: IntegrationInstance):
+    # Drop IPv4 responses
+    assert client.execute("iptables -I INPUT -s 169.254.169.254 -j DROP").ok
+    _test_crawl(client, "http://[fd00:ec2::254]")
+
+    # Block IPv4 requests
+    assert client.execute("iptables -I OUTPUT -d 169.254.169.254 -j REJECT").ok
+    _test_crawl(client, "http://[fd00:ec2::254]")
+
+    # Re-enable IPv4
+    assert client.execute("iptables -D OUTPUT -d 169.254.169.254 -j REJECT").ok
+    assert client.execute("iptables -D INPUT -s 169.254.169.254 -j DROP").ok
+
+    # Drop IPv6 responses
+    assert client.execute("ip6tables -I INPUT -s fd00:ec2::254 -j DROP").ok
+    _test_crawl(client, "http://169.254.169.254")
+
+    # Block IPv6 requests
+    assert client.execute("ip6tables -I OUTPUT -d fd00:ec2::254 -j REJECT").ok
+    _test_crawl(client, "http://169.254.169.254")
```

It'll default all of our ec2 tests to being dual-stack with ipv6 metadata enabled, but I don't see any reason that would be a problem.

File: tests/unittests/sources/test_ec2.py [hunk 7]
```diff
@@ -300,9 +301,14 @@ def register_helper(register, base_url, body):
             register(base_url, "not found", status=404)
 
     def myreg(*argc, **kwargs):
-        url = argc[0]
-        method = httpretty.PUT if ec2.API_TOKEN_ROUTE in url else httpretty.GET
-        return httpretty.register_uri(method, *argc, **kwargs)
+        url, body = argc
+        method = responses.PUT if ec2.API_TOKEN_ROUTE in url else responses.GET
+        print(
```
Comment (TheRealFalcon) [inline tests/unittests/sources/test_ec2.py:306]: This can get rather verbose, so I don't think we wanna keep this around.

File: tests/unittests/sources/test_ec2.py [hunk 8]
```diff
@@ -680,10 +732,11 @@ def test_aws_token_redacted(self):
         logs_with_redacted_ttl = [log for log in all_logs if REDACT_TTL in log]
         logs_with_redacted = [log for log in all_logs if REDACT_TOK in log]
         logs_with_token = [log for log in all_logs if "API-TOKEN" in log]
-        self.assertEqual(1, len(logs_with_redacted_ttl))
-        self.assertEqual(83, len(logs_with_redacted))
+        assert len(logs_with_redacted_ttl)
```
Comment (TheRealFalcon) [inline tests/unittests/sources/test_ec2.py:735]: Do you know why this test changed?

File: cloudinit/url_helper.py [hunk 9]
```diff
@@ -394,31 +517,94 @@ def wait_for_url(
...
+        url = None
+        try:
+            url, response = url_reader_cb(urls)
+            url_exc, reason = read_url_handle_response(response, url)
+            if not url_exc:
+                return (url, response)
+        except UrlError as e:
+            reason = "request error [%s]" % e
+            url_exc = e
+        except Exception as e:  # yikes, do we really think this is necessary?
```
Comment (TheRealFalcon) [inline cloudinit/url_helper.py:567]: Why the yikes? If everything doesn't get caught by UrlError, we'd want a broader exception to catch anything else.

File: cloudinit/url_helper.py [hunk 10]
```diff
@@ -394,31 +517,94 @@ def wait_for_url(
...
         if max_wait is None:
             return False
         return (max_wait <= 0) or (time.time() - start_time > max_wait)
 
-    loop_n = 0
-    response = None
-    while True:
-        if sleep_time_cb is not None:
-            sleep_time = sleep_time_cb(response, loop_n)
+    def read_url_handle_response(response, url):
```
Comment (TheRealFalcon) [inline cloudinit/url_helper.py:530]: Can we rename this to something like `handle_url_response`? On first glance the name looks like it's involved in calling readurl.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
A few minor comments inline, but otherwise looks good. Once we get the dual-stack timeout fixed and the integration test passes, we should be good to go.

Additionally, 2 tickets we should create for follow-up:
1. Add `pytest-responses` to `Build-Depends` of all the packaging branches. I also plan on adding `pytest-mock`, so we could do those at the same time.
2. Replace httpretty with responses everywhere. No need to have 2 libraries doing the same thing.

File: cloudinit/url_helper.py (hunk 9 above)
Comment (holmanb) [reply cloudinit/url_helper.py:567]: Oops. That's a note-to-self that slipped past my final review, agreed.

File: tests/unittests/sources/test_ec2.py (hunk 8 above)
Comment (holmanb) [reply tests/unittests/sources/test_ec2.py]: Reverted

**Review by TheRealFalcon — APPROVED** [review]
This looks good to me now. Thanks for all the work you put into this!
